### PR TITLE
Serialization

### PR DESCRIPTION
**Pull request by sclamons**
This branch:
1) Adds the ability to pickle and unpickle Term, Model, and LineageModel objects. This allows Models/LineageModels to work with copy.copy/deepcopy and be passed directly with multiprocessing.
2) Adds tests for pickling and unpickling Models and LineageModels.
3) Refactors some Term subclasses into subclasses of BinaryTerm(Term), which swallows some redundant code. 
4) Refactors the tests in test_propensities_via_annotations.py into a single parameterized test.

### CODE REVIEW

File: tests/test_serialization.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+		print('\t params =', param_dict)
+
+
+	### DELETEME
+	param_dict["rate"] = "A*3"
+
+	consolidation_rxn = (inputs, outputs, "general", param_dict)
+
+	# We'll add an A-->B reaction with delay.
+	delay_k = round(np.exp(np.random.uniform(low = -1,
```
Comment (WilliamIX) [inline tests/test_serialization.py:181]: Delays are actually not currently supported in lineages due to complications partitioning the delay queue objects, so don't need to include this in the tests.

File: tests/test_serialization.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	M.create_parameter("I0", 10) #Inducer concentration
+	M.create_parameter("T_I0", 2) #Initial time inducer is added
+	M.create_rule("assignment", {"equation":"I = _I0*Heaviside(t-_T_I0)"})
+
+
+	###############################
+	# TEST SERIALIZATION OF MODEL #
+	###############################
+	timepoints = np.arange(0, 25, 0.01)
+	prepickling_results = py_simulate_model(timepoints,
```
Comment (WilliamIX) [inline tests/test_serialization.py:106]: Can we ensure this works with deterministic simulations as well?

**Review by WilliamIX — CHANGES_REQUESTED** [review]
Looks pretty good! two minor requests then I think we can merge.

**Comment by sclamons** [comment]
These make sense to me.

**Comment by sclamons** [comment]
Added deterministic simulation checks alongside the stochastic simulation checks. They're producing some results with NaNs (not entirely NaNs, but some NaNs), but they're _reproducibly_ returning the same results before and after pickling. 

I'm leaving the checks on LineageModels with delay reactions, so that if we ever get around to adding delays these tests will be ready. These tests will still run just fine on LineageModels with delay reactions as long as lineages's behavior is to ignore the delays -- we can't check Model equality directly anyway (that would require a patch's worth of added code), so we can only readily check "equality" by looking at simulation results.

**Review by WilliamIX — APPROVED** [review]

**Pull request merged**